### PR TITLE
Feat/35 implement regex matching

### PR DESCRIPTION
**Pull request by PiaSchroeder**
#### Changes:

- In config: 
	- added regexs 
	- removed active_db
- In db: 
	- removed set_db() and get_db() and all references to it
	- added match_db() --> Function that matches item names with db using regex
	- added db_name parameter to all other functions in db
- In cache: 
	- Adjusted JOB_ID_PATTERN and normalize_name() to also work with the ZensusDB and RegioDB item names
- In helloworld:
	- added db_name to logincheck()
- In http_helper:
	- adjusted JOB_ID_PATTERN to be more specific
	- added db matching to get_data_from_endpoint()
	- added _get_db_from_user_input() for when matching fails
- nb 0_Databases.ipynb might be obsolete?

### CODE REVIEW

File: nb/01_Databases.ipynb [hunk 1]
```diff
@@ -8,67 +8,48 @@
     "\n",
     "Once you have at least once called `setup_credentials()` you are good to use any method of this package with any of the supported databases.\n",
     "\n",
-    "However, before you can do so, you have to tell `pystatis` which database you want to work with. \n",
+    "The respective database (GENESIS-online, Zensusdatenbank, or Regionaldatenbank) will be determined automatically, all you need to provide the item code.\n",
```
Comment (pmayd) [inline nb/01_Databases.ipynb:11]: Here is a verb missing, like "all you need to provide *is* the item code"

File: nb/profile.ipynb [hunk 2]
```diff
@@ -26,7 +26,7 @@
    "outputs": [],
    "source": [
     "# change your password\n",
-    "change_password(new_password=DoNotUseThisAccidentally)\n",
+    "change_password(new_password=\"DoNotUseThisAccidentally\")\n",
```
Comment (pmayd) [inline nb/profile.ipynb:29]: Does this method not need a database argument to function properly?

File: src/pystatis/cache.py [hunk 3]
```diff
@@ -132,7 +132,7 @@ def normalize_name(name: str) -> str:
     Returns:
         str: The unique identifier without the optional job id.
     """
-    if len(re.findall(JOB_ID_PATTERN, name)) == 3:
+    if re.findall(JOB_ID_PATTERN, name):
```
Comment (pmayd) [inline src/pystatis/cache.py:135]: The original idea was to parse a number like `51000-0017_988038177` and omit the last part so that the job id becomes the original evas number. Your code is no longer checking for valid job ids but only for any numbers. I think it still works because `split` just returns the input if it cannot find a "_", but the logic is now different and the pattern is no longer a JOB_ID_PATTERN so we should rename this constant if you did this deliberately

File: src/pystatis/config.py [hunk 4]
```diff
@@ -26,6 +26,9 @@
 PKG_NAME = __name__.split(".", maxsplit=1)[0]
 DEFAULT_CONFIG_DIR = str(Path().home() / f".{PKG_NAME}")
 SUPPORTED_DB = ["genesis", "zensus", "regio"]
+REGEX_DB = ["^((\d{5}-\d{4})|([0-9A-Z]{10}))$",
```
Comment (pmayd) [inline src/pystatis/config.py:29]: I would not use a list here because it is unclear which regex belongs to which database. I would always suggest using a dictionary for that kind of mapping because that is what it is about: mapping a database name to the regex that belongs to it. This is also much more user friendly because you can simply extract the regex by `REGEX_DB.get("genesis")` instead of looking up the index

File: src/pystatis/config.py [hunk 5]
```diff
@@ -124,8 +127,8 @@ def write_config() -> None:
 def create_default_config() -> None:
     """Create a default config parser with empty credentials."""
     config.add_section("settings")
-    config.set("settings", "active_db", "")
     config.set("settings", "supported_db", ",".join(SUPPORTED_DB))
+    config.set("settings", "regex_db", ";".join(REGEX_DB))
```
Comment (pmayd) [inline src/pystatis/config.py:131]: why do we need this in the config?

File: src/pystatis/config.py (hunk 4 above)
Comment (pmayd) [reply src/pystatis/config.py:29]: I think it would be better to use the `re` module here instead of the different places, so I actually would like to store a precompiled regex here, like `re.compile(...)`, this will allow you to later just call `REGEX_DB["genesis"].filterall(<name>)`

File: src/pystatis/db.py [hunk 6]
```diff
@@ -1,61 +1,66 @@
...
-        logger.critical(
-            "No credentials for %s found. Please run `setup_credentials()`.",
-            name,
-        )
+        name (str): Query parameter 'name' corresponding to item code.
 
+    Returns: 
+        db_name (str): Name of matching database. 
+    """
+    supported_dbs = config.get_supported_db()
```
Comment (pmayd) [inline src/pystatis/db.py:20]: you don't need this if you follow my advice to store REGEX_DB as dict

File: src/pystatis/db.py [hunk 7]
```diff
@@ -1,61 +1,66 @@
...
+    if db_match:
+        # If more than one db matches it must be a Cube (provided all regexing works as intended).
+        # --> Choose db based on available credentials.
+        if len(db_match) > 1:
+            for check_db in db_match:
+                if get_db_user(check_db) and get_db_pw(check_db):
+                    return check_db
+            else:
+                raise PystatisConfigError(
+                    "You lack the necessary credentials to access this item. " \
```
Comment (pmayd) [inline src/pystatis/db.py:38]: we should also add the database name here so the user knows which database they are missing

File: src/pystatis/http_helper.py [hunk 8]
```diff
@@ -18,8 +18,8 @@
 
 logger = logging.getLogger(__name__)
 
-JOB_ID_PATTERN = re.compile(r"\d+-\d+_\d+")
-JOB_TIMEOUT = 60
+JOB_ID_PATTERN = re.compile(r"(?<=:\s).*_\d+")
```
Comment (pmayd) [inline src/pystatis/http_helper.py:21]: why was this changed? Could you make a comment or so about the new regex and its meaning?

File: src/pystatis/http_helper.py [hunk 9]
```diff
@@ -92,7 +92,18 @@ def get_data_from_endpoint(
...
+    
+    # Determine database by matching regex to item code
+    name = params.get("name", params.get("selection", ""))
+    use_db = ""
+
+    if name is not None:
+        use_db = db.match_db(name)
+    
+    if not use_db:
+        use_db = _get_db_from_user_input()
```
Comment (pmayd) [inline src/pystatis/http_helper.py:104]: we should not use this here as this would mean in the worst case the method can only be used if the user enters something in the console because we have a `input` call inside this function. Instead, this function should just fail. However, I am not sure how this works out for Find endpoints and things like that where we need to get data but don't have a name/number so I guess we have to think more about this use case

File: src/pystatis/http_helper.py [hunk 10]
```diff
@@ -297,3 +309,21 @@ def _check_destatis_status(destatis_status: dict) -> None:
         logger.info(
             "Code %d: %s", destatis_status_code, destatis_status_content
         )
+
+def _get_db_from_user_input() -> str:
+    """Prompt user to select the database they want to interact with."""
+
+    
+    user_input = input("Please select the database you wish to interact with:\n" \
```
Comment (pmayd) [inline src/pystatis/http_helper.py:317]: we should not hardcode this, we have `supported_db` in the config for this

File: src/pystatis/http_helper.py [hunk 11]
```diff
@@ -92,7 +92,18 @@ def get_data_from_endpoint(
     Returns:
         requests.Response: the response object holding the response from calling the Destatis endpoint.
     """
-    db_host, db_user, db_pw = db.get_db_settings()
+    
+    # Determine database by matching regex to item code
+    name = params.get("name", params.get("selection", ""))
```
Comment (pmayd) [inline src/pystatis/http_helper.py:97]: this can also fail for endpoints like /catalogue/cubes2statistics that accept partial names or /vatalogue/values that has a selection param but that is TODUR1 for example so no EVAS at all. We have to rethink this approach

**Review by pmayd — APPROVED** [review]
I have left a few comments. The main issue I see is that with this new approach it is no longer possible to fetch data from non-data endpoints like catalogue and find without the user having to specify the database name interactively, which might be ok in interactive sessions but not if you use our library in a script, which was also something we considered and should support. I am not sure what the best solution might be in this case. And probably we won't need it either, if no-one is going to use the library like that so it might be ok for the moment. However, I would like to have a proper and good solution instead of the mix we have right now. Maybe we can try to integrate results from all supported databases but that would increase the runtime a lot...

I guess for now we can live with it but we should try to think of a better approach

**Comment by pmayd** [comment]
I just noticed that my objection against the user input is not as strong as I thought because currently, we only use such endpoints in find. And I guess no-one is going to use the Find class in a script or to automate something, this will only be used in an interactive session to search something. Nevertheless, I think the best thing to do would be to specify the database in the Find class and pass it down to the load_data function the same we do it for logincheck etc. So you can either pass a database or if none is given we can try to get the number from the name.

I can implement this idea or leave it to you @PiaSchroeder . I can also fix the tests if you want

File: src/pystatis/db.py [hunk 12]
```diff
@@ -1,61 +1,66 @@
...
+                    "Please run setup_credentials()."
+                )
+        else:
+            return db_match[0]
```
Comment (pmayd) [inline src/pystatis/db.py:42]: just found a small error: When there is only a single match, we return it without checking the credentials. But I believe it would be better to move the check out of this function, anyway. The function should handle matching the db without checking for the credentials, I guess this is a better job for the calling function or another helper method

**Comment by pmayd** [comment]
I fixed all the tests and updated the branch

**Comment by pmayd** [comment]
I implemented a simple support in the Find class for setting the db parameter, so Find uses `db_name` to avoid asking the user, everything else should work as it was

File: src/pystatis/cache.py (hunk 3 above)
Comment (PiaSchroeder) [reply src/pystatis/cache.py:135]: The previous approach only worked for GENESIS table codes because it merely checked if there were three groups of numbers (Regionalstatistik and Zensus tables have letters and varying numbers of number groups). The underscore, however, is specific to the Job ID generated when a table is requested as a background job, so it is safer to check it this way. I am not attached to the name JOB_ID_PATTERN but in my interpretation, the job id pattern is characterized by the underscore and following numbers, so it makes sense to me to keep the name. I am happy to change it though if you disagree.

File: src/pystatis/config.py [hunk 13]
```diff
@@ -178,6 +181,10 @@ def get_supported_db() -> list[str]:
     """Get a list of supported database names."""
     return SUPPORTED_DB
 
+def get_regex() -> list[str]:
```
[the remark this answers is not in the document]
Comment (PiaSchroeder) [reply src/pystatis/config.py:184]: I renamed the functions to ```identify_db``` and ```get_db_identifiers```

File: src/pystatis/http_helper.py (hunk 8 above)
Comment (PiaSchroeder) [reply src/pystatis/http_helper.py:21]: The previously used pattern (```r"\d+-\d+_\d+"```) only matched GENESIS job ids (e.g. 51000-0017_988038177) but not Regionalstatistik (e.g. 12111-31-01-4_123456789) or Zensus (e.g. 1000A-1034_123456789) IDs. The new pattern uses the fact that the genesis API returns a characteristic string when starting a job where the job id follows a colon and white space in the end (positive lookbehind assertion: ```(?<=:\s)```). The pattern ```.*_\d+``` then allows any number and combination of characters that make up the various item codes, followed by an underscore and a group of numbers, which is characteristic of a background job.

File: src/pystatis/config.py (hunk 4 above)
Comment (pmayd) [reply src/pystatis/config.py:29]: @PiaSchroeder this is still relevant

File: src/pystatis/http_helper.py (hunk 9 above)
Comment (pmayd) [reply src/pystatis/http_helper.py]: I already implemented my idea of a workaround or solution for now!

File: src/pystatis/http_helper.py (hunk 11 above)
Comment (pmayd) [reply src/pystatis/http_helper.py]: should be done...either the method gets a db_name or we try to find it in params. If both fails, the method fails

**Comment by pmayd** [comment]
Should be clearer now what is open and what not.
